flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/PendingCheckpointTest.java

```diff
@@ -578,6 +580,55 @@ public void testReportTaskFinishedOperators() throws IOException {
                 contains(ACK_TASKS.get(0).getVertex()));
     }
 
+    /**
+     * We need to trigger disposal prior completing the checkpoint future to avoid a race condition
+     * with the CheckpointCleaner shutdown.
+     */
+    @Test
+    public void testAbortTriggersDisposalPriorCompletingCheckpointFuture() throws Exception {
+        final PendingCheckpoint pending =
+                createPendingCheckpoint(
+                        CheckpointProperties.forCheckpoint(
+                                CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION));
+        final CompletableFuture<Void> completionFuture =
+                pending.getCompletionFuture()
+                        .handle(
+                                (result, error) -> {
+                                    assertNull(result);
+                                    assertNotNull(error);
+                                    assertTrue(pending.isDisposed());
+                                    return null;
+                                });
+        abort(pending, CheckpointFailureReason.CHECKPOINT_DECLINED);
+        completionFuture.get();
+    }
+
+    /**
+     * We need to trigger disposal prior completing the checkpoint future to avoid a race condition
+     * with the CheckpointCleaner shutdown.
+     */
+    @Test
+    public void testFinalizeTriggersDisposalPriorCompletingCheckpointFuture() throws Exception {
+        final PendingCheckpoint pending =
+                createPendingCheckpoint(
+                        CheckpointProperties.forCheckpoint(
+                                CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION));
+        pending.acknowledgeTask(ATTEMPT_ID, null, new CheckpointMetrics(), null);
+        assertTrue(pending.areTasksFullyAcknowledged());
+        final CompletableFuture<Void> completionFuture =
+                pending.getCompletionFuture()
+                        .handle(
+                                (result, error) -> {
+                                    assertNotNull(result);
+                                    assertNull(error);
```

Comment:
same as above

Reply:
This one is already completed exceptionally, but something along the lines of
```
        final CompletableFuture<Void> completionFuture =
                pending.getCompletionFuture()
                        .thenAccept(ignored -> {})
                        .exceptionally(
                                error -> {
                                    assertTrue(pending.isDisposed());
                                    return null;
                                });
```
could work